cpp/src/dictionary/set_keys.cu

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/dictionary/dictionary_column_view.hpp>
+#include <cudf/dictionary/dictionary_factories.hpp>
+#include <cudf/detail/search.hpp>
+#include <cudf/detail/valid_if.cuh>
+#include <cudf/stream_compaction.hpp>
+#include <cudf/detail/stream_compaction.hpp>
+
+#include <rmm/thrust_rmm_allocator.h>
+#include <thrust/binary_search.h>
+
+namespace cudf
+{
+namespace dictionary
+{
+namespace detail
+{
+namespace
+{
+
+/**
+ * @brief Type-dispatch functor for remapping the old indices to new values based on the new key-set.
+ *
+ * The dispatch is based on the key type.
+ * The output column is the new indices column for the new dictionary column.
+ */
+struct dispatch_compute_indices
+{
+    template<typename Element>
+    std::unique_ptr<column> operator()( dictionary_column_view const& input, column_view const& new_keys,
+                                        rmm::mr::device_memory_resource* mr, cudaStream_t stream )
+    {
+        auto dictionary_view = column_device_view::create(input.parent(),stream);
+        auto d_dictionary = *dictionary_view;
+        auto dictionary_itr = thrust::make_transform_iterator( thrust::make_counting_iterator<size_type>(0),
+            [d_dictionary] __device__ (size_type idx) {
+                if( d_dictionary.is_null(idx) )
+                    return Element{};
+                column_device_view d_indices = d_dictionary.child(0);
+                column_device_view d_keys = d_dictionary.child(1);
+                auto index = d_indices.element<int32_t>(idx + d_dictionary.offset());
```

Comment:
Meta issue :  would it make sense to typedef dictionary indices?  To maybe ease transition to 64 bit down the road if it ever came up.